File: android/app/src/main/java/com/lasercars_mobile/SkiaCanvas.java
package com.lasercars_mobile;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Point;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.WindowMetrics;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReadableArray;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class SkiaCanvas extends SurfaceView implements SurfaceHolder.Callback {
    private int width;
    private int height;
    private int backgroundColor;
    private int minLength;
    private int maxLength;
    private int maxLasers;
    private int minSpeed;
    private int speedRange;
    private int minThickness;
    private int maxThickness;
    private int[] colors;

    public SkiaCanvas(Context context) {
        super(context);
        getHolder().addCallback(this);

        Activity activity = getActivity();
        if (activity != null && android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
            WindowMetrics windowMetrics = activity.getWindowManager().getCurrentWindowMetrics();
            // Use windowMetrics
            width = windowMetrics.getBounds().width();
            height = windowMetrics.getBounds().height();
        } else {
            width = Resources.getSystem().getDisplayMetrics().widthPixels;
            height = Resources.getSystem().getDisplayMetrics().heightPixels;
        }
    }

    private Activity getActivity() {
        Context context = getContext();
        while (context instanceof ContextWrapper) {
            if (context instanceof Activity) {
                return (Activity) context;
            }
            context = ((ContextWrapper) context).getBaseContext();
        }
        return null;
    }

    private DrawingThread drawingThread;
    private final List<Laser> lasers = new ArrayList<>();
    private final Random random = new Random();

    @Override
    public void surfaceCreated(@NonNull SurfaceHolder surfaceHolder) {
        drawingThread = new DrawingThread(surfaceHolder, this);
        drawingThread.setRunning(true);
        drawingThread.start();
    }

    @Override
    public void surfaceChanged(@NonNull SurfaceHolder surfaceHolder, int i, int i1, int i2) {
        // Do nothing
    }

    @Override
    public void surfaceDestroyed(@NonNull SurfaceHolder surfaceHolder) {
        boolean retry = true;

        drawingThread.setRunning(false);

        while (retry) {
            try {
                drawingThread.join();
                retry = false;
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public void update(double percentageOfFrame, @NonNull Canvas canvas) {
        canvas.drawColor(backgroundColor);
        for (Laser laser : lasers) {
            laser.update(percentageOfFrame);
            laser.draw(canvas);
        }
    }

    public void SetBackgroundColor(String color) {
        backgroundColor = Color.parseColor(color);
    }

    public void SetMinLength(int minLength) {
        this.minLength = minLength * 2;
        InitIfReady();
    }

    public void SetMaxLength(int maxLength) {
        this.maxLength = maxLength * 2;
        InitIfReady();
    }

    public void SetMaxLasers(int maxLasers) {
        this.maxLasers = maxLasers;
        InitIfReady();
    }

    public void SetLaserColors(@NonNull ReadableArray colors) {
        this.colors = new int[colors.size()];
        for (int i = 0; i < colors.size(); i++) {
            this.colors[i] = Color.parseColor(colors.getString(i));
        }
        InitIfReady();
    }

    public void SetMinSpeed(int minSpeed) {
        this.minSpeed = minSpeed * 2;
        InitIfReady();
    }

    public void SetSpeedRange(int speedRange) {
        this.speedRange = speedRange;
        InitIfReady();
    }

    public void SetMinThickness(int minThickness) {
        this.minThickness = minThickness * 3;
        InitIfReady();
    }

    public void SetMaxThickness(int maxThickness) {
        this.maxThickness = maxThickness * 3;
        InitIfReady();
    }

    private void InitIfReady() {
        if (colors == null || colors.length == 0 || maxLasers <= 0 || minSpeed <= 0 || speedRange <= 0 || minThickness <= 0 || maxThickness <= 0 || minLength <= 0 || maxLength <= 0) {
            return;
        }

        lasers.clear();
        for (int i = 0; i < maxLasers; i++) {
            int length = random.nextInt(maxLength - minLength) + minLength;
            Point start = new Point(random.nextInt(width + height) - length - height, random.nextInt(height + length * 2) - length);
            int speed = random.nextInt(speedRange) + minSpeed;

            lasers.add(new Laser(start, length, speed, colors[random.nextInt(colors.length)]));
        }
    }

    private class Laser {
        Point start;
        int length;
        int speed;
        Paint paint;

        Laser(Point start, int length, int speed, int color) {
            this.start = start;
            this.length = length;
            this.speed = speed;

            paint = new Paint();
            paint.setColor(color);
            paint.setStrokeWidth((float) InterpolateThickness());
        }

        void draw(@NonNull Canvas canvas) {
            canvas.drawLine((float) start.x, (float) start.y, (float) (start.x + length), (float) (start.y + length), paint);
        }

        void update(double percentageOfFrame) {
            start = new Point(start.x + (int) (speed * percentageOfFrame), start.y + (int) (speed * percentageOfFrame));

            if (start.x > width || start.y > height) {
                length = random.nextInt(maxLength - minLength) + minLength;
                start = new Point(random.nextInt(width + height) - length - height, -length);
                paint.setColor(colors[random.nextInt(colors.length)]);
                speed = random.nextInt(speedRange) + minSpeed;
                paint.setStrokeWidth((float) InterpolateThickness());
            }
        }

        public double InterpolateThickness() {
            return ((double) (speed - minSpeed) / (speedRange)) * (maxThickness - minThickness) + minThickness;
        }
    }
}

File: android/app/src/main/java/com/lasercars_mobile/SkiaCanvas.kt
package com.lasercars_mobile

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.content.res.Resources
import android.graphics.*
import android.view.SurfaceHolder
import android.view.SurfaceView
import com.facebook.react.bridge.ReadableArray
import kotlin.random.Random

class SkiaCanvas(context: Context) : SurfaceView(context), SurfaceHolder.Callback {
    private var width: Int
    private var height: Int
    private var backgroundColor: Int = 0
    private var minLength: Int = 0
    private var maxLength: Int = 0
    private var maxLasers: Int = 0
    private var minSpeed: Int = 0
    private var speedRange: Int = 0
    private var minThickness: Int = 0
    private var maxThickness: Int = 0
    private var colors: IntArray? = null

    private var drawingThread: DrawingThread? = null
    private val lasers = ArrayList<Laser>()
    private val random = Random

    init {
        holder.addCallback(this)

        val activity = getActivity()
        if (activity != null && android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
            val windowMetrics = activity.windowManager.currentWindowMetrics
            width = windowMetrics.bounds.width()
            height = windowMetrics.bounds.height()
        } else {
            width = Resources.getSystem().displayMetrics.widthPixels
            height = Resources.getSystem().displayMetrics.heightPixels
        }
    }

    private fun getActivity(): Activity? {
        var context = context
        while (context is ContextWrapper) {
            if (context is Activity) {
                return context
            }
            context = context.baseContext
        }
        return null
    }

    override fun surfaceCreated(surfaceHolder: SurfaceHolder) {
        // Do nothing
    }

    override fun surfaceChanged(surfaceHolder: SurfaceHolder, i: Int, i1: Int, i2: Int) {
        // Do nothing
    }

    override fun surfaceDestroyed(surfaceHolder: SurfaceHolder) {
        var retry = true

        drawingThread?.setRunning(false)

        while (retry) {
            try {
                drawingThread?.join()
                retry = false
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    fun update(percentageOfFrame: Double, canvas: Canvas) {
        canvas.drawColor(backgroundColor)
        for (laser in lasers) {
            laser.update(percentageOfFrame)
            laser.draw(canvas)
        }
    }

    fun setBackgroundColor(color: String) {
        backgroundColor = Color.parseColor(color)
    }

    fun setMinLength(minLength: Int) {
        this.minLength = minLength * 2
        initIfReady()
    }

    fun setMaxLength(maxLength: Int) {
        this.maxLength = maxLength * 2
        initIfReady()
    }

    fun setMaxLasers(maxLasers: Int) {
        this.maxLasers = maxLasers
        initIfReady()
    }

    fun setLaserColors(colors: ReadableArray) {
        this.colors = IntArray(colors.size())
        for (i in 0 until colors.size()) {
            this.colors!![i] = Color.parseColor(colors.getString(i))
        }
        initIfReady()
    }

    fun setMinSpeed(minSpeed: Int) {
        this.minSpeed = minSpeed * 2
        initIfReady()
    }

    fun setSpeedRange(speedRange: Int) {
        this.speedRange = speedRange
        initIfReady()
    }

    fun setMinThickness(minThickness: Int) {
        this.minThickness = minThickness * 3
        initIfReady()
    }

    fun setMaxThickness(maxThickness: Int) {
        this.maxThickness = maxThickness * 3
        initIfReady()
    }

    private fun initIfReady() {
        if (colors == null || colors!!.isEmpty() || maxLasers <= 0 || minSpeed <= 0 || speedRange <= 0 || minThickness <= 0 || maxThickness <= 0 || minLength <= 0 || maxLength <= 0) {
            return
        }

        lasers.clear()
        for (i in 0 until maxLasers) {
            val length = random.nextInt(maxLength - minLength) + minLength
            val start = Point(random.nextInt(width + height) - length - height, random.nextInt(height + length) - length)
            val speed = random.nextInt(speedRange) + minSpeed

            lasers.add(Laser(start, length, speed, colors!![random.nextInt(colors!!.size)]))
        }
    }

    fun setRunning(running: Boolean) {
        if (drawingThread == null) {
            drawingThread = DrawingThread(holder, this)
            drawingThread?.setRunning(running)
            drawingThread?.start()
        }
        if (!running && drawingThread != null) {
            drawingThread?.setRunning(false)
            drawingThread = null
        }
    }

    private inner class Laser(var start: Point, var length: Int, var speed: Int, color: Int) {
        var paint: Paint = Paint()

        init {
            paint.color = color
            paint.strokeWidth = interpolateThickness().toFloat()
        }

        fun draw(canvas: Canvas) {
            canvas.drawLine(start.x.toFloat(), start.y.toFloat(), (start.x + length).toFloat(), (start.y + length).toFloat(), paint)
        }

        fun update(percentageOfFrame: Double) {
            start = Point(start.x + (speed * percentageOfFrame).toInt(), start.y + (speed * percentageOfFrame).toInt())

            if (start.x > width || start.y > height) {
                length = random.nextInt(maxLength - minLength) + minLength
                start = Point(random.nextInt(width + height) - length - height, -length)
                paint.color = colors!![random.nextInt(colors!!.size)]
                speed = random.nextInt(speedRange) + minSpeed
                paint.strokeWidth = interpolateThickness().toFloat()
            }
        }

        private fun interpolateThickness(): Double {
            return ((speed - minSpeed).toDouble() / speedRange) * (maxThickness - minThickness) + minThickness
        }
    }
}